aep/general/0122/aep.md

```diff
@@ -0,0 +1,318 @@
+ID for the publisher, and `les-miserables` is the resource ID for the book.
+
+- Resource IDs **may** be either always set by users (required on resource
+  creation), optionally set by users (optional on resource creation,
+  server-generated if unset), or never set by users (not accepted at resource
+  creation). They **should** be immutable once created.
+  - If resource IDs are user-settable, the API **must** document allowed
+    formats. User-settable resource IDs **should** conform to [RFC-1034][];
+    which restricts to letters, numbers, and hyphen, with the first character a
+    letter, the last a letter or a number, and a 63 character maximum.
+    - Additionally, user-settable resource IDs **should** restrict letters to
+      lower-case (`^[a-z]([a-z0-9-]{0,61}[a-z0-9])?$`).
+    - Characters outside of ASCII **should not** be permitted; however, if
+      Unicode characters are necessary, APIs **must** follow guidance in
+      [AIP-210][].
+    - User-settable IDs **should not** be permitted to be a UUID (or any value
+      that syntactically appears to be a UUID).
+  - If resource IDs are not user-settable, the API **should** document the
+    basic format, and any upper boundaries (for example, "at most 63
+    characters").
+  - For more information, see the [create][] standard method.
+
+[aip-128]: ./0128.md
+[create]: ./0133.md#user-specified-ids
+[rfc-1034]: https://tools.ietf.org/html/rfc1034
+
+### Resource ID aliases
+
+It is sometimes valuable to provide an alias for common lookup patterns for
+resource IDs. For example, an API with `users` at the top of its resource
+hierarchy may wish to provide `users/me` as a shortcut for retrieving
+information for the authenticated user.
+
+APIs **may** provide programmatic aliases for common lookup patterns. However,
+all data returned from the API **must** use the canonical resource path.
+
+### Full resource paths
+
+In most cases, resource paths are used within a single API only, or else they
+are used in contexts where the owning API is clear (for example,
+`string pubsub_topic`).
+
+However, sometimes it is necessary for services to refer to resources in an
+arbitrary API. In this situation, the service **should** use the _full resource
+path_, a schemeless URI with the owning API's service endpoint, followed by the
+relative resource path:
+
+```
+//apis.example.com/library/publishers/123/books/les-miserables
+//apis.example.com/calendar/users/vhugo1802
+```
+
+**Note:** The full resource path **should not** be used for cross-API
+references where the owning API is clear; it is only used if a field refers to
+resources in multiple APIs where ambiguity is possible.
+
+### Resource URIs
+
+The full resource path is a schemeless URI, but slightly distinct from the full
+URIs we use to access a resource. The latter adds two components: the protocol
+(HTTPS) and the API version:
+
+```
+https://apis.example.com/library/v1/publishers/123/books/les-miserables
```

Comment:
this pattern of having a version for an API is harmful to lightweight clients: they can't infer the HTTP path to the resource without a piece of data that's missing in the resource schema (version).

I think this is related to https://github.com/aep-dev/aepc/issues/2, if we are choosing to move design discussions here instead of aepc.

Reply:
Well, I agree that if we were to say that APIs shouldn't be versioned, then this wouldn't make sense.

I don't think, though, that this section is doing anything other than defining/clarifying terms.  The URI for accessing a resource is obviously going to exist, and so what should it be called?  This section answers that.

Reply:
I filed https://github.com/aep-dev/aep.dev/issues/67 for us to track discussing this further. But fair enough that this particular AEP doesn't discuss versioning.

In the future we should hot-link this to the right section that discusses HTTP paths with the conclusion.